### PR TITLE
pluginhandler: patch and handle elf files on glibc mismatch

### PR DESCRIPTION
**Pull request by sergiusens**
If the required glibc is greater than the one provided by the given
base, fetch libc6 from the host, prime it and use the dynamic loader
from there instead.

Closes: #1668 #1669

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

File: snapcraft/internal/elf.py [hunk 1]
```diff
@@ -77,7 +77,6 @@ def __init__(self, *, soname: str, path: str, root_path: str,
             self.in_base_snap = False
 
 
-@lru_cache()
```
Comment (elopio) [inline snapcraft/internal/elf.py]: I'm curious, why did you remove it?

File: snapcraft/internal/pluginhandler/_mangling.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+def _get_dynamic_linker(library_list: List[str]) -> str:
+    """Return the dynamic linker from library_list."""
+    regex = re.compile(r'(?P<dynamic_linker>ld-[\d.]+.so)$')
+
+    for library in library_list:
+        m = regex.search(os.path.basename(library))
+        if m:
+            return m.group('dynamic_linker')
+
+    raise NotImplementedError(
```
Comment (elopio) [inline snapcraft/internal/pluginhandler/_mangling.py]: I don't think this should be a NotImplementedError.
A more snapcraft-specific exception would be better.

Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_mangling.py]: Agreed, although I think this is the one @sergiusens specifically did not want to make friendly. A RuntimeError would be better-suited, if needed.

File: snapcraft/internal/pluginhandler/_mangling.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
```
Comment (kyrofa) [inline snapcraft/internal/pluginhandler/_mangling.py]: We already have a `snapcraft.internal.mangling` module. Should this go in there?

File: snapcraft/internal/pluginhandler/__init__.py [hunk 4]
```diff
@@ -575,7 +550,20 @@ def prime(self, force=False) -> None:
...
+        # get_core_dynamic_linker to resolve the final path for which
+        # we resort to our only working base 16, ld-2.23.so.
+        linker_compatible = (e.is_linker_compatible(linker='ld-2.23.so')
+                             for e in elf_files)
+        if not all((x for x in linker_compatible)):
+            handle_glibc_mismatch(elf_files=elf_files,
+                                  root_path=self.primedir,
+                                  snap_base_path=self._snap_base_path,
+                                  core_base_path=core_path)
+        elif self._confinement == 'classic':
```
Comment (kyrofa) [inline snapcraft/internal/pluginhandler/__init__.py]: Why are we patching the mismatched versions only on non-classic snaps?

File: snapcraft/internal/pluginhandler/_mangling.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        if required_glibc:
+            formatted_list.append('- {} -> GLIBC {}'.format(
+                elf_file.path, required_glibc))
+            patch_elf_files.append(elf_file)
+    logger.warning('The primed files will not work with the current '
+                   'base given the GLIBC mismatch of the primed '
+                   'files and the linker version (2.23) used in the '
+                   'base. These are the GLIBC versions required by '
+                   'the primed files that do not match and will be '
+                   'patched:\n {}\n'.format(
```
Comment (kyrofa) [inline snapcraft/internal/pluginhandler/_mangling.py]: This message implies (to me) that we're patching the mismatching files to use the older version of libc, which isn't the case. We're migrating the newer libc into the snap and patching them to use that instead of the one in the core snap, right? Perhaps this warning could say something like:

"The primed files will not work[...] These are the GLIBC versions required by the primed files that don't match[...]. To work around this, the newer libc will be migrated into the snap, and these files will be patched to use it."

File: snapcraft/internal/elf.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/internal/elf.py]: simple pimple; on running `load_dependencies` to get the new ones, we'd get the old ones. There is no API to wipe the cache so I just decided to remove it

File: snapcraft/internal/pluginhandler/_mangling.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/internal/pluginhandler/_mangling.py]: `RuntimeError` it shall be, this is not anything a user can fix and a horrible thing if it happens so the error should look just as grotesque for us to fix fast.

File: snapcraft/internal/mangling.py [hunk 6]
```diff
@@ -46,3 +52,89 @@ def rewrite_python_shebangs(root_dir):
...
+
+    logger.warning('The primed files will not work with the current '
+                   'base given the GLIBC mismatch of the primed '
+                   'files and the linker version (2.23) used in the '
+                   'base. These are the GLIBC versions required by '
+                   'the primed files that do not match and will be '
+                   'patched:\n {}\n'
+                   'To work around this, the newer libc will be '
+                   'migrated into the snap, and these files will be '
+                   'patched to use it.'.format('\n'.join(formatted_list)))
```
Comment (kyrofa) [inline snapcraft/internal/mangling.py]: This was tremendously helpful when testing, by the way.

File: snapcraft/tests/bin/readelf/readelf [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env python3
```
Comment (kyrofa) [inline snapcraft/tests/bin/readelf/readelf]: Nice idea.

File: snapcraft/tests/unit/test_mangling.py [hunk 8]
```diff
@@ -106,3 +108,116 @@ def test_following_docstring_no_rewrite(self):
             This is a test
             =======================
         """)))
+
+
+class FakeElf(fixtures.Fixture):
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_mangling.py]: We have snapcraft/tests/fixture_setup.py-- should this go in there?

Comment (sergiusens) [reply snapcraft/tests/unit/test_mangling.py]: the only reason I didn't move it into fixtures_setup is because it is so big and I believe it has the wrong name.

I'll propose a followup with a refactor to `fixtures_setup`
